stl/inc/functional

```diff
@@ -1467,6 +1494,23 @@ inline void __stdcall _Function_deallocate_large_default_aligned(_Move_only_func
     ::operator delete(_Self._Large_fn_ptr<void>());
 }
 
+template <class _Fn>
+void __stdcall _Function_destroy_old_large(_Move_only_function_data& _Self) noexcept {
+    _Self._Large_fn_ptr<_Fn>()->_Delete_this(true);
+}
+
+template <class _Fn>
+void __stdcall _Function_destroy_old_small_as_large(_Move_only_function_data& _Self) noexcept {
+    _Fn* const _Old_fn_impl = _Self._Large_fn_ptr<_Fn>();
+    _Old_fn_impl->_Delete_this(false);
+    ::operator delete(static_cast<void*>(_Old_fn_impl));

```

Comment:
The reason why "_small_as_large" exists as a separate move is that we avoid inner `_Deallocate` of `_Delete_this` and free the memory with direct `::operator delete` paired with direct `::operator new`.

We could call `_Allocate` to match `_Deallocate`, but we can't call it with exactly matching size and alignment, because we don't have that information, it was type-erased.

Reply:
It may also feel like we defeated the purpose of inner allocator of `function`, and also likely have to allocate more memory than we need.

But remember that before this optimization we did exactly the same: `move_only_function` does not have inner allocators, and would have allocated the whole `sizeof(std::function)`.

So that's the scenario when this optimization is not useful to avoid allocation, but it doesn't make things any worse either, and the extra invoke is still avoided.

Reply:
Another view on this situation is that old `std::function` implementation should not have taken that `bool` in `_Delete_this`. It could know if it is small or large at compile time. The deleting path for small functions is essentially dead code. 

(`move_only_function` is superior here: it uses different implementations of deleters, including no deleter at all for trivially-destructible small functions and generic deallocator for trivially-destructible and not overaligned large functions (shared for all `move_only_function`s and all callables). It both improves runtime performance and reduces code bloat)